analyses/collapse-rnaseq/cwl/scripts/liftover_collapse_rnaseq.py

```diff
@@ -0,0 +1,177 @@
+    """
+    Checks if entry is in dup dict (really symbol seen > 1 in gene_sym_ct_dict),
+    then add to dict ot process, or output  to open file handle as-is if not repeated
+    """
+    data = in_data.rstrip('\n').split('\t')
+    gene_sym = data[n_idx]
+    out_data = None
+    if gene_sym_ct_dict[gene_sym] > 1:
+        if gene_sym not in dup_gene_sym_dict:
+            dup_gene_sym_dict[gene_sym] = []
+            ct += 1
+        # cast as int for later
+        dup_gene_sym_dict[gene_sym].append(list(map(float, data[d_idx:])))
+    else:
+        out_data = "{}\t{}\n".format(gene_sym, '\t'.join(data[d_idx:]))
+    return ct, out_data
+
+
+def collapse_dups(dup_gene_sym_dict):
+    """
+    Get highest mean for repeated gene symbols and output that row
+    """
+    for gene_sym in dup_gene_sym_dict:
+        means = []
+        for data in dup_gene_sym_dict[gene_sym]:
+            means.append(mean(data))
+        top_idx = means.index(max(means))
+        out_data = "{}\t{}\n".format(gene_sym, '\t'.join(list(map(str, dup_gene_sym_dict[gene_sym][top_idx]))))
+        return out_data
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Script to use a GTF and expression matrix with ENSG to liftover gene symbols."
+    )
+    parser.add_argument(
+        "-t",
+        "--table",
+        action="store",
+        dest="table",
+        help="Input expression matrix. Can be gzipped or plain tsv",
+    )
+    parser.add_argument(
+        "-o", "--output-basename", action="store", dest="out", help="output basename of liftover and collapse if flag given"
+    )
+    parser.add_argument(
+        "-i", "--input-gtf", action="store", dest="gtf", help="GTF file to use as liftover reference. Can be gzipped or plain tsv"
+    )
+    parser.add_argument(
+        "-g", "--gene-id", action="store", dest="gene_id", help="NAME of gene ID (ENSG values) column"
+    )
+    parser.add_argument(
+        "-n", "--gene-name", action="store", dest="gene_name", help="NAME of gene name column, if present to keep if not found"
+    )
+    parser.add_argument(
+        "-s", "--skip", action="store", dest="skip", type=int, help="Number of lines to skip if needed"
+    )
+    parser.add_argument(
+        "-c", "--collapse", action="store_true", dest="collapse", help="If set, will collapse on gene_symbol"
+    )
+
+    args = parser.parse_args()
+
+    gtf_dict = {}
+    # hardcoded for now, could make opts
+    tx_type = 2
+    tx_desc = 8
+    # set up pattern matcher ahead of time for efficiency
+    id_find = re.compile(r"gene_id \"(ENSG\d+)\.\d+\";")
+    name_find = re.compile(r".*gene_name \"(\S+)\";")
+    print('Indexing gtf', file=sys.stderr)
+    with (gzip.open if args.gtf.endswith("gz") else open)(args.gtf, "rt", encoding="utf-8") as gtf:
+        for line in gtf:
+            if not line.startswith('##'):
+                update_gtf_dict(gtf_dict, line, tx_type, tx_desc, id_find, name_find)
```

Comment:
```suggestion
            newline = gtf.readline()
            while newline.startswith('##'):
                newline = gtf.readline()
            while newline != '':
                update_gtf_dict(gtf_dict, newline, tx_type, tx_desc, id_find, name_find)
                newline = gtf.readline()
```
I wonder if this would speed things up. Basically once you go through the header and then stop checking for header lines. GTF files end up being pretty long so those excess checks will add up in runtime.